### PR TITLE
use dependent functions everywhere

### PR DESCRIPTION
**Pull request by eparejatobes**

### CODE REVIEW

File: src/main/scala/cosas/fns/functions.scala [hunk 1]
```diff
@@ -1,61 +1,43 @@
...
-trait Fn0 extends AnyFn0
-
-
-trait AnyFn1 extends AnyFn {
+/* Dependent functions aka dependent products */
+trait AnyDepFn {
+  type Out
+}
+trait AnyDepFn0 extends AnyDepFn
+trait DepFn0[O] extends AnyDepFn0 {
```
Comment (laughedelic) [inline src/main/scala/cosas/fns/functions.scala]: I think better not to have name clashes with shapeless. Let's name them just `FnX` or `DepFunX`?

File: src/main/scala/cosas/fns/functions.scala [hunk 2]
```diff
@@ -117,8 +99,21 @@ trait AnyAt2 extends Any with AnyAt {
   def apply(in1: X1, in2: X2): Y
 }
 
+trait AnyAt3 extends Any with AnyAt {
+
+  type DepFn <: AnyDepFn3
+  type X1 <: DepFn#In1; type X2 <: DepFn#In2; type X3 <: DepFn#In3
+
+  def apply(in1: X1, in2: X2, in3: X3): Y
+}
```
Comment (laughedelic) [inline src/main/scala/cosas/fns/functions.scala]: So in the end what's the point of these traits? I think that having only `AppX` classes is just enough.

File: src/main/scala/cosas/fns/package.scala [hunk 3]
```diff
@@ -17,10 +19,16 @@ package object fns {
...
+    DF0 <: AnyDepFn2,
+    X10 <: DF0#In1, X20 <: DF0#In2,
+    O0 <: DF0#Out
+  ](df: DF0): syntax.DepFn2ApplySyntax[DF0,X10,X20,O0] =
+    syntax.DepFn2ApplySyntax(df)
 
+  implicit def depFn3ApplySyntax[
+    DF0 <: AnyDepFn3,
+    X10 <: DF0#In1, X20 <: DF0#In2, X30 <: DF0#In3,
+    O0 <: DF0#Out
```
Comment (laughedelic) [inline src/main/scala/cosas/fns/package.scala]: oh, please :pray: O0OooO... I think you overuse zeros...

File: src/main/scala/cosas/package.scala [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package ohnosequences
+
+package object cosas {
+
+  sealed trait <:!<[A, B]
```
Comment (laughedelic) [inline src/main/scala/cosas/package.scala]: maybe better to name it `>:<`? :grin:

File: src/main/scala/cosas/package.scala [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+  implicit def nsub[A, B] : A <:!< B = new <:!<[A, B] {}
+  implicit def nsubAmbig1[A, B >: A] : A <:!< B = ???
+  implicit def nsubAmbig2[A, B >: A] : A <:!< B = ???
+
+  case object TypeRef
+  case class TypeRef[X](val x: TypeRef.type) extends AnyVal
+
+  def q[X]: TypeRef[X] = TypeRef(TypeRef)
```
Comment (laughedelic) [inline src/main/scala/cosas/package.scala]: why `q`?

File: src/main/scala/cosas/properties/parsers.scala [hunk 6]
```diff
@@ -5,14 +5,11 @@ import ohnosequences.cosas.types._
...
 
-  type Type = P
-  type Value = V
-}
+case class PropertyParser[P <: AnyProperty,V](
+  val _tpe: P,
+  val _labelRep: String
+)(
+  val _parser: V => Option[P#Raw]
+) extends DenotationParser[P,P#Raw,V](_tpe,_labelRep)(_parser) with AnyPropertyParser
```
Comment (laughedelic) [inline src/main/scala/cosas/properties/parsers.scala]: this piece of code looks like python.. :laughing:

File: src/main/scala/cosas/package.scala (hunk 5 above)
Comment (eparejatobes) [inline src/main/scala/cosas/package.scala]: random choice. I'm happy with any notational choice here

File: src/main/scala/cosas/package.scala (hunk 4 above)
Comment (eparejatobes) [inline src/main/scala/cosas/package.scala]: `!<`

Comment (eparejatobes) [inline src/main/scala/cosas/package.scala]: I also added `!=`

File: src/main/scala/cosas/fns/functions.scala (hunk 2 above)
Comment (eparejatobes) [inline src/main/scala/cosas/fns/functions.scala]: scope issues

File: src/main/scala/cosas/fns/functions.scala (hunk 1 above)
Comment (eparejatobes) [inline src/main/scala/cosas/fns/functions.scala]: fine

File: src/main/scala/cosas/package.scala (hunk 5 above)
Comment (laughedelic) [inline src/main/scala/cosas/package.scala]: `typeRef[X]`?

File: src/main/scala/cosas/fns/functions.scala (hunk 2 above)
Comment (laughedelic) [inline src/main/scala/cosas/fns/functions.scala]: I don't know what you mean, but if it's absolutely necessary, name them at least consistently, i.e `AnyAppX`

File: src/main/scala/cosas/package.scala (hunk 5 above)
Comment (eparejatobes) [inline src/main/scala/cosas/package.scala]: `ofType[X]`

File: src/main/scala/cosas/fns/functions.scala (hunk 2 above)
Comment (eparejatobes) [inline src/main/scala/cosas/fns/functions.scala]: OK

**Comment by eparejatobes** [comment]
OK merging. Any lingering issues will be fixed elsewhere.

**Pull request merged**